tests/Doctrine/Tests/DBAL/Functional/StatementTest.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+namespace Doctrine\Tests\DBAL\Functional;
+
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Types\Type;
+
+class StatementTest extends \Doctrine\Tests\DbalFunctionalTestCase
+{
+    public function testStatementIsReusableAfterClosingCursor()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_reusable');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_reusable', array('id' => 1));
+        $this->_conn->insert('stmt_test_reusable', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_reusable ORDER BY id');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+
+        $stmt->closeCursor();
+
+        $stmt->execute();
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(2, $id);
+    }
+
+    public function testClosedCursorDoesNotContainResults()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_no_results');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_no_results', array('id' => 1));
+        $this->_conn->insert('stmt_test_no_results', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_no_results');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertNotEmpty($id);
+
+        $stmt->closeCursor();
+
+        try {
+            $value = $stmt->fetchColumn();
+        } catch (\Exception $e) {
+            // some adapters trigger PHP error or throw adapter-specific exception in case of fetching
+            // from a closed cursor, which still proves that it has been closed
```

Comment:
@deeky666 if I understand correctly, there's no clear expectation of how cursors should behave after being closed given the current implementation. The problem with returning `false` is that it's hard to identify why exactly fetch failed on the application end. "Fetch Failure" and "Commands out of sync;" are too generic and may mean some other exception which we don't want to silently convert to `false`.

Another question is, do we really have to fix consistency as part of this ticket. It's already big enough and judging by the number resolves ~5 different issues (the number of test cases).